packages/loader/container-loader/src/test/layerCompatValidation.spec.ts

```diff
@@ -0,0 +1,295 @@
+					),
+				"Loader should be compatible with Runtime layer",
+			);
+		});
+	});
+
+	/**
+	 * These tests validates that the Loader layer compatibility is correctly enforced during container creation.
+	 */
+	describe("Container create", () => {
+		function getCodeLoader(compatibilityDetails?: ILayerCompatDetails): ICodeDetailsLoader {
+			return {
+				load: async () => {
+					return {
+						details: {
+							package: "none",
+						},
+						module: {
+							fluidExport: {
+								IRuntimeFactory: {
+									get IRuntimeFactory(): IRuntimeFactory {
+										return this;
+									},
+									async instantiateRuntime(context, existing): Promise<IRuntime> {
+										return failSometimeProxy<IRuntime & IProvideLayerCompatDetails>({
+											createSummary: () => ({
+												tree: {},
+												type: SummaryType.Tree,
+											}),
+											setAttachState: () => {},
+											getPendingLocalState: () => ({
+												pending: [],
+											}),
+											ILayerCompatDetails: compatibilityDetails,
+										});
+									},
+								},
+							},
+						},
+					};
+				},
+			} satisfies ICodeDetailsLoader;
+		}
+
+		it("Older Runtime is compatible", async () => {
+			const loader = new Loader({
+				codeLoader: getCodeLoader(),
+				documentServiceFactory: failProxy(),
+				urlResolver: failProxy(),
+			});
+			await assert.doesNotReject(
+				async () => loader.createDetachedContainer({ package: "none" }),
+				"Older Runtime should be compatible",
+			);
+		});
+
+		it("Runtime with generation >= minSupportedGeneration is compatible", async () => {
+			const runtimeCompatDetails: ILayerCompatDetails = {
+				pkgVersion,
+				generation: RuntimeSupportRequirements.minSupportedGeneration,
+				supportedFeatures: new Set(),
+			};
+			const loader = new Loader({
+				codeLoader: getCodeLoader(runtimeCompatDetails),
+				documentServiceFactory: failProxy(),
+				urlResolver: failProxy(),
+			});
+
+			await assert.doesNotReject(
+				async () => loader.createDetachedContainer({ package: "none" }),
+				"Runtime with generation >= minSupportedGeneration should be compatible",
+			);
+		});
+
+		it("Runtime with generation < minSupportedGeneration is not compatible", async () => {
+			const runtimeGeneration = RuntimeSupportRequirements.minSupportedGeneration - 1;
+			const runtimeCompatDetails: ILayerCompatDetails = {
+				pkgVersion,
+				generation: runtimeGeneration,
+				supportedFeatures: new Set(),
+			};
+			const loader = new Loader({
+				codeLoader: getCodeLoader(runtimeCompatDetails),
+				documentServiceFactory: failProxy(),
+				urlResolver: failProxy(),
+			});
+
+			await assert.rejects(
+				async () => loader.createDetachedContainer({ package: "none" }),
+				(error: Error) =>
+					validateFailureProperties(
+						error,
+						false /* isGenerationCompatible */,
+						runtimeGeneration,
+					),
+				"Runtime with generation >= minSupportedGeneration should be incompatible",
+			);
+		});
+	});
```

Comment:
If you want to add coverage that disposeFn is called.  `Sinon.fake` is a nice utility, you can do `Sinon.fake()` to get a no-op fn, or you can do `Sinon.fake.returns(42)` or `Sinon.fake.throws(new Error("foo"))` or a few other things too.

```suggestion
		const disposeFn = Sinon.fake();

		assert.throws(
			() => validateRuntimeCompatibility(runtimeCompatDetails, disposeFn),
			(e: Error) =>
				validateFailureProperties(
					e,
					false /* isGenerationCompatible */,
					runtimeGeneration,
					requiredFeatures,
				),
			"Loader should be compatible with Runtime layer",
		);
		assert(disposeFn.calledOnce, "Dispose function should be called");
```

Reply:
(also fixed typo in assert msg)